components/script/dom/htmlimageelement.rs

```diff
@@ -428,23 +411,55 @@ impl HTMLImageElement {
         let _ = task_source.queue(task, window.upcast());
     }
 
-    fn set_current_request_url_to_selected(&self, src: DOMString) {
-        struct SetUrlToSelectedTask {
+    /// Step 11.4 of https://html.spec.whatwg.org/multipage/#update-the-image-data
+    fn set_current_request_url_to_selected_fire_error_and_loadend(&self, src: DOMString) {
+        struct Task {
             img: Trusted<HTMLImageElement>,
             src: String,
         }
-        impl Runnable for SetUrlToSelectedTask {
+        impl Runnable for Task {
+            fn handler(self: Box<Self>) {
+                // Step 9.2
+                let img = self.img.root();
+                let mut current_request = img.current_request.borrow_mut();
+                current_request.source_url = Some(DOMString::from_string(self.src));
+                img.upcast::<EventTarget>().fire_event(atom!("error"));
+                img.upcast::<EventTarget>().fire_event(atom!("loadend"));
+            }
+        }
+
+        let task = box Task {
+            img: Trusted::new(self),
+            src: src.into()
+        };
+        let document = document_from_node(self);
+        let window = document.window();
+        let task_source = window.dom_manipulation_task_source();
+        let _ = task_source.queue(task, window.upcast());
+    }
+
+    fn set_current_request_url_to_string_and_fire_load(&self, src: DOMString, url: ServoUrl) {
+        struct SetUrlToStringTask {
+            img: Trusted<HTMLImageElement>,
+            src: String,
```

Comment:
Storing a DOMString gives the error `the trait `core::marker::Send` is not implemented for `dom::bindings::str::DOMString`, also it seems DOMString has `!Send` implemented (I tried adding an empty `Send` implementation, which gave me a `conflicting implementation` error...